pilot/pkg/model/push_context.go

```diff
@@ -1707,12 +1707,26 @@ func (ps *PushContext) BestEffortInferServiceMTLSMode(service *Service, port *Po
 		return MTLSUnknown
 	}
 
-	// First , check mTLS settings from beta policy (i.e PeerAuthentication) at namespace / mesh level.
+	// 1. Check service instances' tls mode, mainly used for headless service.
+	if service.Resolution == Passthrough {
+		instances := ps.InstancesByPort(service, port.Port, nil)
+		if len(instances) == 0 {
+			return MTLSDisable
+		}
+		for _, i := range instances {
+			// Infer mTls disabled if any of the endpoint is with tls disabled
```

Comment:
With normal eds we select mtls per endpoint. This gives us correct behavior in the presence of mixed endpoints. I think we could do the same with headless in theory - we can create a tls and non tls cluster and configure the per-endpoint listeners? wdyt?

Reply:
i have thought about this idea a while ago. Yeah in theory that would work.

By the way, how today's outbound traffic is routed onto headless outbound cluster? We have one listener per TCP service port? Or we already have the per-endpoint outbound listener already?

Reply:
The concern might be whether push performance & latency w.r.t the endpoint change rate. But if we already have the outbound listener per endpoint, might not be an issue.

Reply:
For tcp headless service, we generate separate listeners. In theory, we can generate one cluster per listener. But i think there is a rare case, the instances are different.

Reply:
We don't really need 1 cluster per listener right? We need at most two clusters: 1 for TLS and one for non TLS

Reply:
That's true. The only obstacle lies on  the cluster name generation, we may need to update everywhere. And we have to add    calling to `push.InstancesByPort`.

Reply:
Talked with @incfly, this will be a followup.

Reply:
https://github.com/istio/istio/issues/27376